File: pkg/metricscollector/v1beta1/common/pns.py
# Copyright 2022 The Kubeflow Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import psutil
import sys
import time
import const


def WaitMainProcesses(pool_interval, timout, wait_all, completed_marked_dir):
    """
    Hold metrics collector parser until required pids are finished
    """

    if not sys.platform.startswith('linux'):
        raise Exception("Platform '{}' unsupported".format(sys.platform))

    pids, main_pid = GetMainProcesses(completed_marked_dir)

    return WaitPIDs(pids, main_pid, pool_interval, timout, wait_all, completed_marked_dir)


def GetMainProcesses(completed_marked_dir):
    """
    Return array with all running processes pids and main process pid which metrics collector is waiting.
    """
    pids = set()
    main_pid = 0
    this_pid = psutil.Process().pid

    for proc in psutil.process_iter():
        # Get pid from process object
        pid = proc.pid

        # Get parent process
        ppid = proc.ppid()

        # Ignore the pause container, our own pid, and non-root processes (parent pid != 0)
        if pid == 1 or pid == this_pid or ppid != 0:
            continue

        # Read the process command line, join all cmdlines in one string
        cmd_lind = " ".join(proc.cmdline())

        # By default main_pid is the first process.
        # In addition to that, command line contains completed marker for the main pid.
        # For example: echo completed > /var/log/katib/$$$$.pid
        # completed_marked_dir is the directory for completed marker, e.g. /var/log/katib
        if main_pid == 0 or ("echo {} > {}".format(const.TRAINING_COMPLETED, completed_marked_dir) in cmd_lind):
            main_pid = pid

        pids.add(pid)

    # If mainPid has not been found, return an error.
    if main_pid == 0:
        raise Exception("Unable to find main pid from the process list {}".format(pids))

    return pids, main_pid


def WaitPIDs(pids, main_pid, pool_interval, timout, wait_all, completed_marked_dir):
    """
    Waits until all pids are finished.
    If waitAll == false WaitPIDs waits until main process is finished.
    """
    start = 0
    # not_finished_pids contains pids that are not finished yet
    not_finished_pids = set(pids)

    if pool_interval <= 0:
        raise Exception("Poll interval seconds must be a positive integer")

    # Start main wait loop
    # We should exit when timeout is out or not_finished_pids is empty
    while (timout <= 0 or start < timout) and len(not_finished_pids) > 0:
        finished_pids = set()
        for pid in not_finished_pids:
            # If pid is completed /proc/<pid> dir doesn't exist
            path = "/proc/{}".format(pid)
            if not os.path.exists(path):
                if pid == main_pid:
                    # For main_pid we check if file with "completed" marker exists if completed_marked_dir is set
                    if completed_marked_dir:
                        mark_file = os.path.join(completed_marked_dir, "{}.pid".format(pid))
                        # Check if file contains "completed" marker
                        with open(mark_file) as file_obj:
                            contents = file_obj.read()
                            if contents.strip() != const.TRAINING_COMPLETED:
                                raise Exception(
                                    "Unable to find marker: {} in file: {} with contents: {} for pid: {}".format(
                                        const.TRAINING_COMPLETED, mark_file, contents, pid))
                    # Add main pid to finished pids set
                    finished_pids.add(pid)
                    # Exit loop if wait all is false because main pid is finished
                    if not wait_all:
                        return
                # Add not main pid to finished pids set
                else:
                    finished_pids.add(pid)

        # Update not finished pids set with finished_pids set
        not_finished_pids = not_finished_pids - finished_pids
        # Sleep for pool_interval seconds before next attempt
        time.sleep(pool_interval)
        start = start + pool_interval

    # After main loop not finished pids set should be empty
    if not_finished_pids:
        raise Exception("Timed out waiting for pids to complete")

File: pkg/metricscollector/v1beta1/common/pns.go
/*
Copyright 2022 The Kubeflow Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package common

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	psutil "github.com/shirou/gopsutil/v3/process"
	"k8s.io/klog"
)

// WaitPidsOpts is the input options for metrics collector
type WaitPidsOpts struct {
	PollInterval           time.Duration
	Timeout                time.Duration
	WaitAll                bool
	CompletedMarkedDirPath string
}

// WaitMainProcesses holds metrics collector parser until required pids are finished.
func WaitMainProcesses(opts WaitPidsOpts) error {

	if runtime.GOOS != "linux" {
		return fmt.Errorf("platform '%s' unsupported", runtime.GOOS)
	}

	pids, mainPid, err := GetMainProcesses(opts.CompletedMarkedDirPath)
	if err != nil {
		return err
	}

	return WaitPIDs(pids, mainPid, opts)
}

// GetMainProcesses returns array with all running processes pids
// and main process pid which metrics collector is waiting.
func GetMainProcesses(completedMarkedDirPath string) (map[int]bool, int, error) {
	pids := make(map[int]bool)
	allPids, err := psutil.Pids()
	mainPid := 0

	if err != nil {
		return nil, 0, fmt.Errorf("failed to list processes: %v", err)
	}

	thisPID := os.Getpid()
	for _, pid := range allPids {
		// Create process object from pid
		proc, err := psutil.NewProcess(pid)
		if err != nil {
			klog.Infof("Unable to create new process from pid: %v, error: %v. Continue to next pid", pid, err)
			continue
		}

		// Get parent process
		ppid, err := proc.Ppid()
		if err != nil {
			klog.Infof("Unable to get parent process for pid: %v, error: %v. Continue to next pid", pid, err)
			continue
		}

		// Ignore the pause container, our own pid, and non-root processes (parent pid != 0)
		if pid == 1 || pid == int32(thisPID) || ppid != 0 {
			continue
		}

		// Read the process command line
		cmdline, err := proc.Cmdline()
		if err != nil {
			klog.Infof("Unable to get cmdline from pid: %v, error: %v. Continue to next pid", pid, err)
			continue
		}

		// By default mainPid is the first process.
		// In addition to that, command line contains completed marker for the main pid
		// For example: echo completed > /var/log/katib/$$$$.pid
		// completedMarkedDirPath is the directory for completed marker, e.g. /var/log/katib
		if mainPid == 0 ||
			strings.Contains(cmdline, fmt.Sprintf("echo %s > %s", TrainingCompleted, completedMarkedDirPath)) {
			mainPid = int(pid)
		}

		pids[int(pid)] = true
	}

	// If mainPid has not been found, return an error.
	if mainPid == 0 {
		return nil, 0, fmt.Errorf("unable to find main pid from the process list %v", allPids)
	}

	return pids, mainPid, nil
}

// WaitPIDs waits until all pids are finished.
// If waitAll == false WaitPIDs waits until main process is finished.
func WaitPIDs(pids map[int]bool, mainPid int, opts WaitPidsOpts) error {

	// notFinishedPids contains pids that are not finished yet
	notFinishedPids := pids

	// Get info from options
	waitAll := opts.WaitAll
	timeout := opts.Timeout
	endTime := time.Now().Add(timeout)
	pollInterval := opts.PollInterval

	// Start main wait loop
	// We should exit when timeout is out or notFinishedPids is empty
	for (timeout == 0 || time.Now().Before(endTime)) && len(notFinishedPids) > 0 {
		// Start loop over not finished pids
		for pid := range notFinishedPids {
			// If pid is completed /proc/<pid> dir doesn't exist
			path := fmt.Sprintf("/proc/%d", pid)
			_, err := os.Stat(path)
			if err != nil {
				if os.IsNotExist(err) {
					if pid == mainPid {
						// For mainPid we check if file with "completed" marker exists if CompletedMarkedDirPath is set
						if opts.CompletedMarkedDirPath != "" {
							markFile := filepath.Join(opts.CompletedMarkedDirPath, fmt.Sprintf("%d.pid", pid))
							// Read file with "completed" marker
							contents, err := os.ReadFile(markFile)
							if err != nil {
								return fmt.Errorf("training container is failed. Unable to read file %v for pid %v, error: %v", markFile, pid, err)
							}
							// Check if file contains "early-stopped" marker
							// In that case process is not completed
							if strings.TrimSpace(string(contents)) == TrainingEarlyStopped {
								continue
							}
							// Check if file contains "completed" marker
							if strings.TrimSpace(string(contents)) != TrainingCompleted {
								return fmt.Errorf("unable to find marker: %v in file: %v with contents: %v for pid: %v",
									TrainingCompleted, markFile, string(contents), pid)
							}
						}
						// Delete main pid from map with pids
						delete(notFinishedPids, pid)
						// Exit loop if wait all is false because main pid is finished
						if !waitAll {
							return nil
						}
						// Delete not main pid from map with pids
					} else {
						delete(notFinishedPids, pid)
					}
					// We should receive only not exist error when we check /proc/<pid> dir
				} else {
					return fmt.Errorf("fail to check process info: %v", err)
				}
			}
		}
		// Sleep for pollInterval seconds before next attempt
		time.Sleep(pollInterval)
	}

	// After main loop notFinishedPids map should be empty
	if len(notFinishedPids) != 0 {
		return fmt.Errorf("timed out waiting for pids to complete")
	}
	return nil
}
